### PR TITLE
Introduce OV Extension base api

### PR DESCRIPTION
**Pull request by ilyachur**
### Details:
 - Introduced base `ov::Extension` API
     - `ov::OpExtension` is needed to register custom operations
 - Implemented new template extension library
     - Library contains custom `Identity` operation
 - Extended `ov::runtime::Core` API
     - `add_extension("path_to_library")`
     - `add_extension(std::vector<ov::Extension::Ptr>&)`
     - `add_extension<Extension type, ...>(ext, ...)`
     - `add_extension<Operation op, ...>()`
 - New API supports old and new extensions to read models with custom operations
 - Added `add_extensions(std::vector<ov::Extension::Ptr>&)` to `Frontend` API
 - Changed `IR Frontend`:
    - Support of new extensions
    - removed `use_framework_node`
- Developed new tests to check that network with custom operation can be read with old and new extensions in new API 

### Tickets:
 - 58738

### CODE REVIEW

**Comment by ilyachur** [comment]
The replacement for: https://github.com/openvinotoolkit/openvino/pull/7204

File: ngraph/core/include/openvino/core/extension.hpp [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+#    define OPENVINO_EXTENSION_C_API extern "C" __declspec(dllexport)
+#    define OPENVINO_EXTENSION_API   __declspec(dllexport)
+#else
+#    define OPENVINO_EXTENSION_C_API extern "C" OPENVINO_API
+#    define OPENVINO_EXTENSION_API   OPENVINO_API
+#endif
+
+namespace ov {
+
+class OPENVINO_API Extension {
```
Comment (apankratovantonp) [inline ngraph/core/include/openvino/core/extension.hpp:25]: enable shared_from_this

File: ngraph/core/include/openvino/core/extension.hpp [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+public:
+    using Ptr = std::shared_ptr<Extension>;
+    virtual ~Extension();
```
Comment (apankratovantonp) [inline ngraph/core/include/openvino/core/extension.hpp:28]: make destructor protected?

File: ngraph/core/include/openvino/core/extension.hpp (hunk 1 above)
Comment (ilyachur) [reply ngraph/core/include/openvino/core/extension.hpp:25]: Why do we need to have shared_from_this for extensions?

Comment (apankratovantonp) [reply ngraph/core/include/openvino/core/extension.hpp:25]: It is intended to be used just with `std::shared_ptr` is not it? So it is just trivial optimization to place reference_cointer inside Extension class and reduce heap fragmentation a little bit in cases when it is created not by std::make_shared.

File: docs/template_extension/new/op.cpp [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+using namespace TemplateExtension;
+
+//! [op:ctor]
+Operation::Operation(const ngraph::Output<ngraph::Node>& arg, int64_t add) : Op({arg}), add(add) {
+    constructor_validate_and_infer_types();
+}
+//! [op:ctor]
+
+//! [op:validate]
+void Operation::validate_and_infer_types() {
```
Comment (apankratovantonp) [inline docs/template_extension/new/op.cpp:16]: I think we need a more complex example operation to show how implement all overloaded methods

Comment (ilyachur) [reply docs/template_extension/new/op.cpp:16]: It is a copy/past of existed operation

File: docs/template_extension/new/op.cpp [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    size_t size = ngraph::shape_size(arg0->get_shape());
+    implementation(arg0->get_data_ptr<ET>(), out->get_data_ptr<ET>(), add, size);
+    return true;
+}
+
+}  // namespace
+
+bool Operation::evaluate(const ngraph::HostTensorVector& outputs, const ngraph::HostTensorVector& inputs) const {
+    switch (inputs[0]->get_element_type()) {
+    case ngraph::element::Type_t::i8:
```
Comment (apankratovantonp) [inline docs/template_extension/new/op.cpp:60]: We need some helper function to reduce such `switch-case` bloat. For example in `arm-plugin` https://github.com/openvinotoolkit/openvino_contrib/blob/bd192a7a1cca7d580f5446dbb8473e2229807fff/modules/arm_plugin/src/arm_converter/arm_converter_gather.cpp#L43 
But it uses c++14

**Comment by apankratovantonp** [comment]
Just want to add. The current solution is not only inconvenient for users but also UNSAFE. Due to exceptions, the `unload` function won't be called even by the user. I strongly recommend moving `so` management to extension containing classes before merging PR!

**Comment by ilyachur** [comment]
> Just want to add. The current solution is not only inconvenient for users but also UNSAFE. Due to exceptions, the `unload` function won't be called even by the user. I strongly recommend moving `so` management to extension containing classes before merging PR!

In general user cannot get an access to extensions from shared library. Only our components work with these extensions.
Any way, I added internal SOExtension class to control the process of destroying extension objects.

**Comment by ilyachur** [comment]
@apankratovantonp @ilya-lavrenov As I said, I changed the mechanism and added SOExtension.
It allows to avoid unload functions.
So I merge this PR.
If you have any other comments I will fix their in the separate PR.

File: ngraph/core/src/so_extension.hpp [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    SOExtension(const std::shared_ptr<void>& so, const Extension::Ptr& ext) : m_so(so), m_ext(ext) {}
+
+    const Extension::Ptr& extension() const;
+
+    const std::shared_ptr<void> shared_object() const;
+
+private:
+    std::shared_ptr<void> m_so;
+    Extension::Ptr m_ext;
+};
```
Comment (ilya-lavrenov) [inline ngraph/core/src/so_extension.hpp:25]: can we just move `SoPtr` from IE to `ov::utils` and reuse here?

**Pull request merged**